File: common/src/main/java/au/com/codeka/warworlds/common/XmlIterator.java
package au.com.codeka.warworlds.common;

import org.w3c.dom.Element;
import org.w3c.dom.Node;

import java.util.Iterator;

/**
 * This is an implementation of {@link Iterator} which lets us more easily iterator over
 * Java XML nodes.
 */
public class XmlIterator {

  /**
   * This iterator iterates child elements of the given parent element.
   */
  public static Iterable<Element> childElements(Element elem) {
    return childElements(elem, null);
  }

  /**
   * This iterator iterates child elements of the given parent element, with the given node name.
   */
  public static Iterable<Element> childElements(Element elem, final String name) {
    Node node;
    for (node = elem.getFirstChild(); node != null; node = node.getNextSibling()) {
      if (node.getNodeType() == Node.ELEMENT_NODE &&
          (name == null || node.getNodeName().equals(name))) {
        break;
      }
    }
    final Node firstNode = node;

    return new Iterable<Element>() {
      @Override
      public Iterator<Element> iterator() {
        return new Iterator<Element>() {
          Node childNode = firstNode;

          private Node findNext(Node startingNode) {
            for (Node n = startingNode.getNextSibling(); n != null; n = n.getNextSibling()) {
              if (n.getNodeType() == Node.ELEMENT_NODE &&
                  (name == null || n.getNodeName().equals(name))) {
                return n;
              }
            }

            return null;
          }

          @Override
          public boolean hasNext() {
            return childNode != null;
          }

          @Override
          public Element next() {
            Element nextElement = (Element) childNode;
            childNode = findNext(childNode);
            return nextElement;
          }

          @Override
          public void remove() {
            childNode = findNext(childNode);
          }
        };
      }
    };
  }
}

File: common/src/main/kotlin/au/com/codeka/warworlds/common/XmlIterator.kt
package au.com.codeka.warworlds.common

import org.w3c.dom.Element
import org.w3c.dom.Node

/**
 * This is an implementation of [Iterator] which lets us more easily iterator over
 * Java XML nodes.
 */
object XmlIterator {
  /**
   * This iterator iterates child elements of the given parent element.
   */
  fun childElements(elem: Element, name: String? = null): Iterable<Element> {
    var node: Node?
    node = elem.firstChild
    while (node != null) {
      if (node.nodeType == Node.ELEMENT_NODE &&
          (name == null || node.nodeName == name)) {
        break
      }
      node = node.nextSibling
    }
    val firstNode = node
    return object : Iterable<Element> {
      override fun iterator(): Iterator<Element> {
        return object : Iterator<Element> {
          var childNode = firstNode
          private fun findNext(startingNode: Node?): Node? {
            var n = startingNode!!.nextSibling
            while (n != null) {
              if (n.nodeType == Node.ELEMENT_NODE &&
                  (name == null || n.nodeName == name)) {
                return n
              }
              n = n.nextSibling
            }
            return null
          }

          override fun hasNext(): Boolean {
            return childNode != null
          }

          override fun next(): Element {
            val nextElement = childNode as Element
            childNode = findNext(childNode)
            return nextElement
          }
        }
      }
    }
  }
}